tests/test_base.py

```diff
@@ -0,0 +1,133 @@
+        )
+        output = step.fit_transform(self.data)
+        assert output == self.data['input_1']
+
+        step = Step(
+            name='test_inputs_without_conflicting_names_do_not_require_adapter_2',
+            transformer=Dummy(),
+            input_data=['input_1', 'input_2'],
+            cache_dirpath=CACHE_DIRPATH
+        )
+        output = step.fit_transform(self.data)
+        assert output == {**self.data['input_1'], **self.data['input_2']}
+
+    def test_inputs_with_conflicting_names_require_adapter(self):
+        step = Step(
+            name='test_inputs_with_conflicting_names_require_adapter',
+            transformer=Dummy(),
+            input_data=['input_1', 'input_3'],
+            cache_dirpath=CACHE_DIRPATH
+        )
+        with self.assertRaises(StepsError):
+            step.fit_transform(self.data)
+
+    def test_adapter_creates_defined_keys(self):
+        step = Step(
+            name='test_adapter_creates_defined_keys',
+            transformer=Dummy(),
+            input_data=['input_1', 'input_2'],
+            adapter={'X': [('input_1', 'features')],
+                     'Y': [('input_2', 'extra_features')]},
+            cache_dirpath=CACHE_DIRPATH
+        )
+        res = step.fit_transform(self.data)
+        self.assertEqual({'X', 'Y'}, set(res.keys()))
+
+    def test_adapter_recipe_with_single_item(self):
+        step = Step(
+            name='test_adapter_recipe_with_single_item',
+            transformer=Dummy(),
+            input_data=['input_1'],
+            adapter={'X': ('input_1', 'features')},
+            cache_dirpath=CACHE_DIRPATH
+        )
+        res = step.fit_transform(self.data)
+        self.assertTrue(np.array_equal(res['X'], self.data['input_1']['features']))
+
+    def test_adapter_recipe_with_list(self):
+        step = Step(
+            name='test_adapter_recipe_with_list',
+            transformer=Dummy(),
+            input_data=['input_1', 'input_2'],
+            adapter={'X': [],
+                     'Y': [('input_1', 'features')],
+                     'Z': [('input_1', 'features'), ('input_2', 'extra_features')]},
+            cache_dirpath=CACHE_DIRPATH
+        )
+        res = step.fit_transform(self.data)
+        for i, key in enumerate(('X', 'Y', 'Z')):
+            self.assertIsInstance(res[key], list)
+            self.assertEqual(len(res[key]), i)
+
+        self.assertEqual(res['X'], [])
+
+        self.assertTrue(np.array_equal(res['Y'][0], self.data['input_1']['features']))
+
+        self.assertTrue(np.array_equal(res['Z'][0], self.data['input_1']['features']))
+        self.assertTrue(np.array_equal(res['Z'][1], self.data['input_2']['extra_features']))
+
+    def test_adapter_recipe_with_list_and_function(self):
+        step = Step(
+            name='test_adapter_recipe_with_list_and_function',
+            transformer=Dummy(),
+            input_data=['input_1', 'input_2'],
+            adapter={'X': ([('input_1', 'features'),
```

Comment:
Personally I think this kind of adapter is hard to understand. My preference would be that if you want to concatenate features from different input steps then you should just create a concatenation step.

Reply:
First, let me notice that the syntax for this case hasn't changed :)

In this PR I executed my proposition described in issue #28 thread. My idea was to retain as much of existing behaviour as possible, while:
1. Simplifying syntax for the simplest and most popular case (removing brackets if you just want one item).
2. Making the syntax more consistent (recipe containing a list yields a list of objects).

If you have ideas for different approach to adapting inputs present them in issue #28 :)